test/e2e/kubelet_stats.go

```diff
@@ -172,6 +177,143 @@ func HighLatencyKubeletOperations(c *client.Client, threshold time.Duration, nod
 	return badMetrics, nil
 }
 
+// getContainerInfo contacts kubelet for the conatiner informaton. The "Stats"
+// in the returned ContainerInfo is subject to the requirements in statsRequest.
+func getContainerInfo(c *client.Client, nodeName string, req *kubelet.StatsRequest) (map[string]cadvisor.ContainerInfo, error) {
+	reqBody, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	data, err := c.Post().
+		Prefix("proxy").
+		Resource("nodes").
+		Name(fmt.Sprintf("%v:%v", nodeName, ports.KubeletPort)).
+		Suffix("stats/container").
+		SetHeader("Content-Type", "application/json").
+		Body(reqBody).
+		Do().Raw()
+
+	var containers map[string]cadvisor.ContainerInfo
+	err = json.Unmarshal(data, &containers)
+	if err != nil {
+		return nil, err
+	}
+	return containers, nil
+}
+
+const (
+	// cadvisor records stats about every second.
+	cadvisorStatsPollingIntervalInSeconds float64 = 1.0
+	// cadvisor caches up to 2 minutes of stats (configured by kubelet).
+	maxNumStatsToRequest int = 120
+)
+
+// A list of containers for which we want to collect resource usage.
+var targetContainers = []string{
+	"/",
+	"/docker-daemon",
+	"/kubelet",
+	"/kube-proxy",
+	"/system",
+}
+
+type containerResourceUsage struct {
+	Name                    string
+	Timestamp               time.Time
+	CPUUsageInCores         float64
+	MemoryUsageInBytes      int64
+	MemoryWorkingSetInBytes int64
+	// The interval used to calculate CPUUsageInCores.
+	CPUIntervalInSeconds float64
+}
+
+// getOneTimeResourceUsageOnNode queries the node's /stats/container endpoint
+// and returns the resource usage of targetContainers for the past
+// cpuInterval.
+// The acceptable range of the interval is 2s~120s. Be warned that as the
+// interval (and #containers) increases, the size of kubelet's response
+// could be sigificant. E.g., the 60s interval stats for ~20 containers is
+// ~1.5MB. Don't hammer the node with frequent, heavy requests.
+// TODO: Implmenet a constant, lightweight resource monitor, which polls
+// kubelet every few second, stores the data, and reports meaningful statistics
+// numbers over a longer period (e.g., max/mean cpu usage in the last hour).
+//
+// cadvisor records cumulative cpu usage in nanoseconds, so we need to have two
+// stats points to compute the cpu usage over the interval. Assuming cadvisor
+// polls every second, we'd need to get N stats points for N-second interval.
+// Note that this is an approximation and may not be accurate, hence we also
+// write the actual interval used for calcuation (based on the timestampes of
+// the stats points in containerResourceUsage.CPUIntervalInSeconds.
+func getOneTimeResourceUsageOnNode(c *client.Client, nodeName string, cpuInterval time.Duration) (map[string]*containerResourceUsage, error) {
+	numStats := int(float64(cpuInterval.Seconds()) / cadvisorStatsPollingIntervalInSeconds)
```

Comment:
Since the client/test is usually running on another machine, clock skew (or simply inconsistent clocks) would affect the StartTime. IMO, StartTime would be more useful in the case where more than one request is sent (and we can tracked the timestamps recorded in previous stats).

Another way to calculate a more accurate interval is to retrieve all 120 stats, scan the timestamps, and get the expected interval (which may not always succeed). This seems a bit more complicated than what this one-time resource usage PR intends to achieve. I would implement this if more accuracy is required (but again, my followup PR does periodic polling for this purpose).